MobileWallet/TariLib/Tor/OnionManager.swift

```diff
@@ -0,0 +1,296 @@
+
+            let args = torConf.arguments!
+
+            #if DEBUG
+            dump("\n\n\(String(describing: args))\n\n")
+            #endif
+            torConf.arguments = args
+            self.torThread = TorThread(configuration: torConf)
+            needsReconfiguration = false
+
+            self.torThread!.start()
+
+            print("[\(String(describing: OnionManager.self))] Starting Tor")
+        } else {
+            if needsReconfiguration {
+                // Not using bridges, so null out the "Bridge" conf
+                torController!.setConfForKey("usebridges", withValue: "0", completion: { _, _ in
+                })
+                torController!.resetConf(forKey: "bridge", completion: { _, _ in
+                })
+            }
+        }
+
+        // Wait long enough for tor itself to have started. It's OK to wait for this
+        // because Tor is already trying to connect; this is just the part that polls for
+        // progress.
+        DispatchQueue.main.asyncAfter(deadline: .now() + 1, execute: {
+            if OnionManager.TOR_LOGGING {
+                // Show Tor log in iOS' app log.
+                TORInstallTorLogging()
+                TORInstallEventLogging()
+            }
+
+            if !(self.torController?.isConnected ?? false) {
+                do {
+                    try self.torController?.connect()
+                } catch {
+                    print("[\(String(describing: OnionManager.self))] error=\(error)")
+                }
+            }
+
+            self.torController?.authenticate(with: OnionManager.cookie, completion: { success, _ in
+                if success {
+                    var completeObs: Any?
+                    completeObs = self.torController?.addObserver(forCircuitEstablished: { established in
+                        if established {
+                            self.state = .connected
+                            self.torController?.removeObserver(completeObs)
+                            self.cancelInitRetry()
+                            self.cancelFailGuard()
+                            #if DEBUG
+                            print("[\(String(describing: OnionManager.self))] connection established")
+                            #endif
+
+                            self.torController?.getSessionConfiguration({ configuration in
+                                delegate?.torConnFinished(configuration: configuration!)
+                            })
+                        }
+                    }) // torController.addObserver
+                    var progressObs: Any?
+                    progressObs = self.torController?.addObserver(forStatusEvents: {
+                        (type: String, _: String, action: String, arguments: [String: String]?) -> Bool in
+
+                        if type == "STATUS_CLIENT" && action == "BOOTSTRAP" {
+                            let progress = Int(arguments!["PROGRESS"]!)!
```

Comment:
we should use optional here because otherwise the app will crash if on of these are nil.